File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque 
def solution(maps):
    m = len(maps)
    n = len(maps[0])
    queue = deque()
    queue.append([0,0,1])
    maps[0][0] = -1 
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    
    while queue:
        x,y,move = queue.popleft()
        
        if x == n-1 and y == m-1:
            return move
        
        for i in range(4):
            mx = dx[i] + x
            my = dy[i] + y 
            if 0<= mx <= n-1 and  0<= my <= m-1:
                if maps[my][mx] == 1:
                    queue.append([mx,my,move+1])
                    maps[my][mx] = -1 
            
        
        
    answer = -1
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        
        int answer = 0;
        int N = maps.length; 
        int M = maps[0].length;
        
        int[] dx = {0,0,-1,1};
        int[] dy = {-1,1,0,0};
        Queue<List<Integer>> q = new LinkedList<>();
        q.offer(Arrays.asList(0,0));

        while (!q.isEmpty()){
            
            List<Integer> now = q.poll();
            
            for (int i = 0; i<4; i++){
                
                int mx = now.get(0) + dx[i];
                int my = now.get(1) + dy[i];
                
                if ( 0<= mx && mx< M && 0<= my && my< N && maps[my][mx] == 1){
                    
                    q.offer(Arrays.asList(mx,my));
                    maps[my][mx] = maps[now.get(1)][now.get(0)] + 1; 

                }
            }   
        }        
        
        if (maps[N-1][M-1] == 1){
            answer = -1;
        }
        else{
            answer = maps[N-1][M-1];
        }
        
        return answer;
    }
}